radixdlt/src/main/java/com/radixdlt/consensus/liveness/WeightedRotatingLeaders.java

```diff
@@ -0,0 +1,142 @@
+ * views closest to the highest view calculated.
+ *
+ * This class stateful and is NOT thread-safe.
+ */
+public final class WeightedRotatingLeaders implements ProposerElection {
+
+	private final ValidatorSet validatorSet;
+	private final Comparator<Entry<Validator, UInt256>> weightsComparator;
+	private final CachingNextLeaderComputer nextLeaderComputer;
+
+	public WeightedRotatingLeaders(ValidatorSet validatorSet, Comparator<Validator> comparator, int cacheSize) {
+		this.validatorSet = validatorSet;
+		this.weightsComparator = Comparator
+			.comparing(Entry<Validator, UInt256>::getValue)
+			.thenComparing(Entry::getKey, comparator);
+		this.nextLeaderComputer = new CachingNextLeaderComputer(validatorSet, weightsComparator, cacheSize);
+	}
+
+	private static class CachingNextLeaderComputer {
+		private final ValidatorSet validatorSet;
+		private final Comparator<Entry<Validator, UInt256>> weightsComparator;
+		private final Map<Validator, UInt256> weights;
+		private final Validator[] cache;
+		private View curView;
+
+		private CachingNextLeaderComputer(ValidatorSet validatorSet, Comparator<Entry<Validator, UInt256>> weightsComparator, int cacheSize) {
+			this.validatorSet = validatorSet;
+			this.weightsComparator = weightsComparator;
+			this.weights = new HashMap<>();
+			this.cache = new Validator[cacheSize];
+			this.resetToView(View.of(0));
+		}
+
+		private Validator computeHeaviest() {
+			final Entry<Validator, UInt256> max = weights.entrySet().stream()
+				.max(weightsComparator)
+				.orElseThrow(() -> new IllegalStateException("Weights cannot be empty"));
+			return max.getKey();
+		}
+
+		private void computeNext() {
+			// Reset last leader
+			final int curIndex = (int) (this.curView.number() % cache.length);
+			final Validator curLeader = cache[curIndex];
+			weights.merge(curLeader, validatorSet.getTotalPower(), UInt256::subtract);
+
+			// Add weight
+			for (Validator validator : validatorSet.getValidators()) {
+				weights.merge(validator, validator.getPower(), UInt256::add);
+			}
+
+			// Compute next leader
+			this.curView = this.curView.next();
+			int index = (int) (this.curView.number() % cache.length);
+			cache[index] = computeHeaviest();
+		}
+
+		private Validator checkCacheForProposer(View view) {
+			if (view.compareTo(curView) <= 0 && view.number() > curView.number() - cache.length) {
+				return cache[(int) (view.number() % cache.length)];
+			}
+
+			return null;
+		}
+
+		private void computeToView(View view) {
+			while (view.compareTo(curView) > 0) {
+				computeNext();
+			}
+		}
+
+		private Validator resetToView(View view) {
+			// reset if view isn't in cache
+			if (curView == null || view.number() < curView.number() - cache.length) {
+				curView = View.of(0);
+				for (Validator validator : validatorSet.getValidators()) {
+					weights.put(validator, UInt256.from(UInt128.ONE, UInt128.ZERO).subtract(validator.getPower()));
```

Comment:
(if so, need to test that assumption here, or maybe even change the data type of `Validator.power`)

Reply:
assuming validator.getPower() < 2^128. This is part of ValidatorSet contract but gonna change this so that power is just 128 bits so it becomes compile time check